rbc/heavydb/geo_nested_array.py

```diff
@@ -0,0 +1,336 @@
+    )
+
+    def impl(geo, lst):
+        nrows = len(lst)
+        geo_ptr = as_voidptr(get_alloca(geo))
+
+        indices = get_c_ptr(nrows, nb_types.int64)
+        n_elems = 0
+        for i in range(nrows):
+            n_elems += len(lst[i])
+            indices[i] = len(lst[i])
+
+        data = get_c_ptr(n_elems, nb_types.double)
+        idx = 0
+        for i in range(nrows):
+            ncols = len(lst[i])
+            for j in range(ncols):
+                data[idx] = lst[i][j]
+                idx += 1
+
+        return fromCoords(geo_ptr, data, indices, nrows)
+
+    return impl
+
+
+def heavydb_geo_fromCoords_vec3(geo, lst):
+    base_type = geo.base_type
+    fromCoords = external(
+        f"void {base_type}_fromCoords(int8_t*, double*, int64_t*, int64_t*, int64_t, int64_t)|cpu"
+    )
+
+    def impl(geo, lst):
+        geo_ptr = as_voidptr(get_alloca(geo))
+
+        indices_i = []
+        indices_j = []
+        flatten = []
+
+        n_matrices = len(lst)
+        for i in range(n_matrices):
+            cnt = 0
+            nrows = len(lst[i])
+            for j in range(nrows):
+                ncols = len(lst[i][j])
+                cnt += ncols
+                indices_j.append(ncols)
+                for k in range(ncols):
+                    flatten.append(lst[i][j][k])
+            indices_i.append(cnt)
+
+        data = get_c_ptr(len(flatten), nb_types.double)
+        for i, e in enumerate(flatten):
+            data[i] = e
+
+        indices_i_ = get_c_ptr(len(indices_i), nb_types.int64)
+        for i, e in enumerate(indices_i):
+            indices_i_[i] = e
+
+        indices_j_ = get_c_ptr(len(indices_j), nb_types.int64)
+        for j, e in enumerate(indices_j):
+            indices_j_[j] = e
+
+        return fromCoords(
+            geo_ptr, data, indices_i_, indices_j_, len(indices_i), len(indices_j)
+        )
+
+    return impl
+
+
+def heavydb_geo_toCoords_vec(geo):
+    base_type = geo.base_type
+    get_value = external(f"double {base_type}_toCoords_get_value(int8_t*, int64_t)|cpu")
+
+    def impl(geo):
+        geo_ptr = as_voidptr(get_alloca(geo))
+
+        lst = []
+        n_coords = geo.n_coords()
+        for i in range(n_coords):
+            lst.append(get_value(geo_ptr, i))
+        return lst
+
+    return impl
+
+
+def heavydb_geo_toCoords_vec2(geo):
+    base_type = geo.base_type
+    get_value = external(
+        f"double {base_type}_toCoords_get_value(int8_t*, int64_t, int64_t)|cpu"
+    )
+
+    def impl(geo):
+        lst = []
+        geo_ptr = as_voidptr(get_alloca(geo))
+        n_rings = geo.n_rings()
+        for i in range(n_rings):
+            n_coords = geo[i].n_coords()
+            inner = []
+            for j in range(n_coords):
+                inner.append(get_value(geo_ptr, i, j))
+            lst.append(inner)
+        return lst
+
+    return impl
+
+
+def heavydb_geo_toCoords_vec3(geo):
+    raise NotImplementedError()
```

Comment:
Could you create an rbc issue about this?

Reply:
Created #543